pennylane/measurements/vn_entropy.py

```diff
@@ -61,4 +64,13 @@ def circuit_entropy(x):
     .. seealso:: :func:`pennylane.qinfo.transforms.vn_entropy` and :func:`pennylane.math.vn_entropy`
     """
     wires = Wires(wires)
-    return MeasurementProcess(VnEntropy, wires=wires, log_base=log_base)
+    return _VnEntropy(VnEntropy, wires=wires, log_base=log_base)
+
+
+class _VnEntropy(StateMeasurement):
+    """Measurement process that returns the Von Neumann entropy."""
+
+    def process_state(self, state: Sequence[complex], wires: Wires):
+        return qml.math.vn_entropy(
+            state, indices=self.wires, c_dtype="complex128", base=self.log_base
```

Comment:
Again, we should preserve the dtype of the state.

Reply:
And add a test using a different `c_dtype` for the device.

Reply:
Changed to `state.dtype`.

I realised that right now `DefaultQubit.state.dtype` is always `np.complex128` (even if you change the `c_dtype` of the device). When integrating this functionality with the device, we could either cast the state to the `self.c_dtype` before passing it into the measurement, or cast the result of the measurement to `self.c_dtype`.

What do you think?

Reply:
That's strange 🤔 . We should be casting in the [`device.state`](https://github.com/PennyLaneAI/pennylane/blob/f1bf5441a5b704da412a01a5b26cdf42279fee3e/pennylane/devices/default_qubit.py#L663) property, but we don't. We already do the casting in [`device.density_matrix`](https://github.com/PennyLaneAI/pennylane/blob/f1bf5441a5b704da412a01a5b26cdf42279fee3e/pennylane/_qubit_device.py#L1172) so there shouldn't be any issues.

I think we can leave this as `state.dtype` for now, and change the `device.state` property in a new PR to do the casting.